docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,753 @@
+     */
+    Optional<GetObjectRequest> objectRequest();
+}
+
+/**
+ * Progress listener for a Transfer.
+ * <p>
+ * The SDK guarantees that calls to {@link #transferProgressEvent(EventContext)}
+ * are externally synchronized.
+ */
+public interface TransferProgressListener {
+    /**
+     * Called when a new progress event is available for a Transfer.
+     *
+     * @param ctx The context object for the given transfer event.
+     */
+    void transferProgressEvent(EventContext ctx);
+
+    interface EventContext {
+    }
+
+    interface Initiated extends EventContext {
+        /**
+         * The transfer this listener associated with.
+         */
+        Transfer<?> transfer();
+    }
+
+    interface BytesTransferred extends Initiated {
+        /**
+         * Average number of bytes transferred per second for this transfer.
+         */
+        long averageTransferBytesPerSecond();
+
+        /**
+         * The transfer request for the object whose bytes were transferred.
+         */
+        TransferObjectRequest objectRequest();
+
+        /**
+         * The number of bytes transferred for this event.
+         */
+        long bytes();
+
+        /**
+         * The total length of the object.
+         */
+        long objectLength();
+
+        /**
+         * If the transfer of the given object is complete.
+         */
+        boolean complete();
+    }
+
+    interface Completed extends Initiated {
+    }
+
+    interface Cancelled extends Initiated {
+    }
+
+    interface Failed extends Initiated {
+        /**
+         * The error.
+         */
+        Throwable error();
+    }
+}
+
+/**
+ * Hints to help optimize an object download.
+ */
+public interface DownloadHints {
+    /**
+     * The size of the object to be downloaded. In parallel download situations
+     * this allows the Transfer Manager to avoid making an extra request to S3
+     * to find the size of the object in order to determine proper object
+     * partitioning.
+     */
+    long objectLength();
+}
+
+/**
+ * A download transfer of a single object from S3.
+ */
+public interface DownloadObject extends Transfer<DownloadObject> {
+    @Override
+    CompletableFuture<DownloadObject> completionFuture();
+
+    @Override
+    DownloadObjectState pause();
+}
+
+
+/**
+ * An upload transfer of a single object to S3.
+ */
+public interface UploadObject extends Transfer<UploadObject> {
+    @Override
+    CompletableFuture<UploadObject> completionFuture();
+
+    @Override
+    UploadObjectState pause();
+}
+
+/**
+ * Represents the transfer of one or more objects to or from S3.
+ */
+public interface Transfer<T extends Transfer> {
+
+    CompletableFuture<T> completionFuture();
+
+    /**
+     * Pause this transfer, cancelling any requests in progress.
+     * <p>
+     * The returned state object can be used to resume this transfer at a later
+     * time.
```

Comment:
Can it? Wouldn't the customer need to know the subtype to actually do that?

Reply:
The subclasses of `Transfer` return the correct type, e.g. `DownloadObjectState`

Reply:
Would it be useful to use this base type method?

Reply:
Nope. I think we can remove it and just add it to the concrete implementations